docs(release-highlights): Adds 6.3 highlights. (#4682)
patternfly/patternfly-org: packages/documentation-site/patternfly-docs/content/get-started/release-highlights.md

- Form
    - **Requires opt-in.**
    - **Example:** [Form field groups](/components/forms/form#field-groups), which have opted in.  

- Navigation
    - **Example:** [Expandable navigation examples](/components/navigation#expandable).
    - **Note:** This animation was introduced in our 6.2 release, but performance issues were reported. For this release, we reworked the animation to resolve these issues.

- Search input   
    - **Example:** [Expandable search input](/components/search-input#with-expandable-button).

- Tree view
    - **Requires opt-in.**
    - **Example:** All [tree view](/components/tree-view) examples and demos, which have opted in. 

- Table
    - **Requires opt-in.**
    - **Example:** [Expandable table](/components/table#expandable), which has opted in.
    - **Note:** This release does not include expansion animations for tree tables.

#### Feedback 

The following components contain animations that give users dynamic feedback, helping to communicate that an action has taken or will take place.  

1.  Buttons
    - **Animation:** When a button is clicked, the active state color ripples outward from the center of the button.

>     - **Example:** All buttons in our examples and demos, including our [animated favorite button example](/components/button#favorite).

Reviewer: This doesn't go to the favorite button example (maybe it works when deployed?). Anyway, I think it would be good to separate the button click ripple from the favorite button star pulse. We talk about the favorite button below, so maybe here we just call out the ripple effect when (most) any button is clicked.

Author: oh great point, ty! I don't think I meant to link there

import { Divider, Timestamp } from '@patternfly/react-core'; 

<Timestamp date={new Date(2025, 6)}>July 2025</Timestamp>
## PatternFly 6.3

### Promoted package versions
The following packages were promoted with this release. Outside of our primary initiatives for this release, we addressed reported bugs and continued to enhance PatternFly with new features, as detailed in the linked changelogs. 

- patternfly/patternfly ([changelog](https://github.com/patternfly/patternfly/releases/tag/v6.3.0))
    - [@patternfly/patternfly@6.3.0](https://www.npmjs.com/package/@patternfly/patternfly)
- patternfly/react ([changelog](https://github.com/patternfly/patternfly-react/releases/tag/v6.3.0))
    - [@patternfly/react-charts@8.3.0](https://www.npmjs.com/package/@patternfly/react-charts)
    - [@patternfly/react-code-editor@6.3.0](https://www.npmjs.com/package/@patternfly/react-code-editor)
    - [@patternfly/react-core@6.3.0](https://www.npmjs.com/package/@patternfly/react-core)
    - [@patternfly/react-docs@7.3.0](https://www.npmjs.com/package/@patternfly/react-docs)
    - [@patternfly/react-drag-drop@6.3.0](https://www.npmjs.com/package/@patternfly/react-drag-drop)

>     - [@patternfly/react-icons@6.3.0](https://www.npmjs.com/package/@patternfly/react-icons)

Author: ah, ty! Missed those

We added expansion animations that apply a fade-in and slide-in transition as hidden content sections are expanded in a variety of components. A similar, reversed animation applies when expanded sections are collapsed.

The following list highlights opt-in requirements and links to examples. Unless an animation is marked with "Requires opt-in," it will be enabled by default. 

- Accordion 
    - **Example:** All [accordions](/components/accordion) in our examples and demos.

- Dual list selector 
    - **Requires opt-in.**
    - **Example:** [Dual list selector with tree](/components/dual-list-selector#with-tree), which has opted in.

- Expandable section 
    - **Example:** All [expandable sections](/components/expandable-section) in our examples and demos.

- Form
    - **Requires opt-in.**
    - **Example:** [Form field groups](/components/forms/form#field-groups), which have opted in.  

- Navigation
    - **Example:** [Expandable navigation examples](/components/navigation#expandable).
    - **Note:** This animation was introduced in our 6.2 release, but performance issues were reported. For this release, we reworked the animation to resolve these issues.

- Search input   

>     - **Example:** [Expandable search input](/components/search-input#with-expandable-button).

Reviewer 2: This one has an opt-in (`hasAnimations` on the `expandableInput` object)

```suggestion
    - **Requires opt-in.**
    - **Example:** [Expandable search input](/components/search-input#with-expandable-button), which has opted in.
```

- Tree view
    - **Requires opt-in.**
    - **Example:** All [tree view](/components/tree-view) examples and demos, which have opted in. 

- Table
    - **Requires opt-in.**

>     - **Example:** [Expandable table](/components/table#expandable), which has opted in.

Reviewer 2: Not sure if you wanted to call it out but ["compound expandable"](https://staging.patternfly.org/components/table#compound-expandable) is also animated (when you click on an open row item to close it, then click one to open it again - doesn't animate when you click between items in a row)

    - **Note:** This release does not include expansion animations for tree tables.

#### Feedback 

> 

Author: adding, thanks!

The following components contain animations that give users dynamic feedback, helping to communicate that an action has taken or will take place.  

1.  Buttons
    - **Animation:** When a button is clicked, the active state color ripples outward from the center of the button.
    - **Example:** All buttons in our components, including in our [button examples and demos](/components/button).

1.  Form validation
    - **Animation:** When an error occurs in a form field with validation, the component "jiggles" from side to side as danger styles are applied. 

>     - **Example:** The [validated form select](/components/forms/form-select#validated), [invalid text area](/components/forms/text-area#invalid), and [invalid text input](/components/forms/text-input#invalid) examples.

Reviewer 2: The `<TextInputGroup />` component also animates. It's similar to the text input component you mentioned, but is a separate component.

### Component animations

As we progress in our effort to integrate motion design across PatternFly components, we added the following new animations in this release. To provide a holistic view of our animated components, we also put together [an interactive animations demo](/design-foundations/motion/react-demos), which we will continue to update as we add additional animation support.   

We have also created an [enable-animations codemod](https://github.com/patternfly/pf-codemods/tree/main/packages/eslint-plugin-pf-codemods/src/rules/v6/enableAnimations), which adds the `hasAnimations` prop to components that support opt-in animations.  

#### Expansion 

We added expansion animations that apply a fade-in and slide-in transition as hidden content sections are expanded in a variety of components. A similar, reversed animation applies when expanded sections are collapsed.

The following list highlights opt-in requirements and links to examples. Unless an animation is marked with "Requires opt-in," it will be enabled by default. 

- Accordion 
    - **Example:** All [accordions](/components/accordion) in our examples and demos.

> - Dual list selector 

Reviewer 3: Would we want to specifically call out this is only for DualListSelectors that have `isTree` on this line?

Author: Thanks! Updating to be "Dual list selector with tree", rather than the prop

    - **Requires opt-in.**
    - **Example:** [Dual list selector with tree](/components/dual-list-selector#with-tree), which has opted in.

> - Expandable section 

Reviewer 3: Maybe worth a note that "detached" ExpandableSections will require manual update to pass in a new `direction` prop (similar to how the ExpandableSectionToggle needs `direction` passed in to have the icon rotate in the right direction), so that the content animates in the right direction. Right now if using a detached Expandablesection and `direction` isn't passed in, there'll be no animation.

    - **Example:** All [expandable sections](/components/expandable-section) in our examples and demos.

- Form
    - **Requires opt-in.**
    - **Example:** [Form field groups](/components/forms/form#field-groups), which have opted in.  

- Navigation
    - **Example:** [Expandable navigation examples](/components/navigation#expandable).
    - **Note:** This animation was introduced in our 6.2 release, but performance issues were reported. For this release, we reworked the animation to resolve these issues.

- Search input   
    - **Requires opt-in.**
    - **Example:** [Expandable search input](/components/search-input#with-expandable-button), which has opted in.

- Tree view
    - **Requires opt-in.**
    - **Example:** All [tree view](/components/tree-view) examples and demos, which have opted in. 

> - Table

Reviewer 3: There may be other changes a consumer has to make to get expandable and compound expansion animations to work correclty/look right. [PR 11865](https://github.com/patternfly/patternfly-react/pull/11865/files) is the actual animation part, while [PR 11815](https://github.com/patternfly/patternfly-react/pull/11815/files) helps resolve a visual issue with a divider line not being remoed on expanded rows.

- For the divider aspect, basically the first `Tr` in the `Tbody` needs to be given `isExpanded` logic, similar to what a consumer may be using on the other `Tr` that contains the expandable content. Then, both the `Table` and the `Tr` that contains the expandable content both need isExpandable passed in
- For animations, really more for compund expandable, but the `Tr` that is `isControlRow` passed in also needs `isExpanded` logic to avoid a visual issue with the animations. Then, each expandable content needs to be rendered in its own unique structure (see below) rather than using a single structure to dynamically render the content.

```
<Tr isExpanded={logic here} isExpandable>
  <Td>
    <ExpandableRowContent>
      The actual content here
    </ExpandableRowContent>
  </Td>
</Tr>
```

In our compound expandable exmaple, we used to only render one of those structures ^ then just dynamically put the correct content inside of `<ExpandableRowContent>`. With the animations update, we had to update example/demo to render 3 of those structures (since we were only showing 3 comound expandable things)

Let me know if you think any of it is worth including here and if you need I can try and help workshop some verbiage with you.

Reviewer 3: FWIW the React examples for expandable and compound exampandable should be up-to-date in terms of the expected structure/prop usage as well as having animations enabled. So we could simple link consumers to the compound expandable table for how to set things up for animations to work. Between the two prop usage is probably similar and would be worth highlighting here, just because Table can be so complex to look through the example code it may not be clear immediately. Let me take a look at your updates and see fi there's any other suggestions to make

1.  Menu toggle validation
    - **Animation:** When an error occurs in a menu toggle with validation, the component "jiggles" from side to side as danger styles are applied. 
    - **Example:** The [danger option in the select with validation](/components/menus/select#with-validation) example.

1.  Progress
    - **Animation:** When an error occurs in a progress bar, the component "jiggles" from side to side as danger styles are applied. 
    - **Example:** The "Danger variant" option in our [progress helper text example](/components/progress#helper-text).

1.  Tabs 
    - **Requires opt-in for HTML-only implementations.**
    - **Animation:** The bottom "active tab" line slides between tabs as they are selected.
    - **Example:** All [default tabs](/components/tabs) in our examples and demos. This animation does not apply to boxed tabs.
    - **Note:** To avoid a breaking change, this animation is opt-in for HTML-only implementations. To support this, view our [HTML guidance for animating the tab accent mark](/components/tabs#animate-current-tab-accent).

> ##### Icons

Reviewer 3: Maybe worth calling out, but for these icon animations, currently there's no ability to customize the icon used. So if the consumer is passing isFavorite, isHamburger, or isSettings, they're stuck with the icon for that respective button.

1.  Favorite
    - **Animation:** When a favorite button is clicked, the star "pulses" and fills with color.
    - **Example:** All favorite buttons in our examples and demos, including our [animated favorite button example](/components/button#favorite).

1.  Hamburger menu 
    - **Animation:** The middle bar of the hamburger icon transforms into an arrow to indicate whether the menu will expand or collapse when clicked. The behavior differs slightly for desktop and mobile views:
        - **Desktop:** Both arrows appears on hover or keyboard focus as appropriate.
        - **Mobile:** When the menu is expanded, a "collapse" arrow is displayed. When collapsed, an "expand" arrow appears on hover or focus.

>     - **Example:** All hamburger menu buttons in our examples and demos, including [our page examples](/components/page#vertical-navigation).

Reviewer 3: Would we want to include the verbiage from the hamburger button example that this animation will occur automatically when in a `<Masthead>`, but otherwise the consumer needs to hook up logic with the `hamburgerVariant` property?

Also would we want to link to the actual hamburger button example as well?

    - [@patternfly/react-virtualized-extension@6.1.0](https://www.npmjs.com/package/@patternfly/react-virtualized-extension/v/6.1.0) ([changelog](https://github.com/patternfly/react-virtualized-extension/releases/tag/v6.1.0))

### React 19 support

PatternFly now includes support for React 19 across all components and extensions, enabling you to benefit from  new React features and performance enhancements.

This update is non-breaking, and PatternFly 6 will continue to support React 17 and 18 in line with our support policy. In future releases, we plan to more fully leverage React 19's capabilities to enhance existing PatternFly features and introduce new ones.

### Component animations

As we progress in our effort to integrate motion design across PatternFly components, we added the following new animations in this release. To provide a holistic view of our animated components, we also put together [an interactive animations demo](/design-foundations/motion/react-demos), which we will continue to update as we add additional animation support.   

#### enable-animations codemods

> We have also created an [enable-animations codemod](https://github.com/patternfly/pf-codemods/tree/main/packages/eslint-plugin-pf-codemods/src/rules/v6/enableAnimations), which adds the `hasAnimations` prop to components that support opt-in animations.  

Reviewer: We need to be super clear here that using this codemod to enable these particular animations could cause breaking changes. That's why they are created as opt-in - they create breaking changes due to changes in the HTML structure.

Instead of "support opt-in" perhaps it's also clearer to say "require opt-in"

- Form
    - **Requires opt-in.**
    - **Example:** [Form field groups](/components/forms/form#field-groups), which have opted in.  

- Navigation
    - **Example:** [Expandable navigation examples](/components/navigation#expandable).
    - **Note:** This animation was introduced in our 6.2 release, but performance issues were reported. For this release, we reworked the animation to resolve these issues.

- Search input   
    - **Requires opt-in.**
    - **Example:** [Expandable search input](/components/search-input#with-expandable-button), which has opted in.

- Tree view
    - **Requires opt-in.**
    - **Example:** All [tree view](/components/tree-view) examples and demos, which have opted in. 

- Table
    - **Requires opt-in.**
    - **Example:** [Expandable table](/components/table#expandable) and [compound expandable table](/components/table/#compound-expandable), which have both opted in.
    - **Notes:** 
        - This release does not include expansion animations for tree tables.
        - To prevent visual issues with the divider on expanded rows, ensure that both the `<Table>` and the first `Tr` in the `Tbody` contains `isExpanded` logic. 

>         - To avoid visual issues with animations in compound expandable tables, ensure that the `Tr` which contains `isControlRow` contains `isExpanded` logic. To dynamically render the content, each expandable row content should also be rendered in its own structure. 

Reviewer 3: I think what you have here is a great start for all of this. Some comments that would be worth tweaking/adding so consumers are structuring their tables in the way that was originally intended by Core but wasn't done so in React (but is now in expandable-like examples), moreso if they're opting into animations but they should also update their structure regardless (but nothing should break if they don't and aren't opting into animations):

- `Table` must have `isExpandable` passed to it (rather than `isExpanded` logic)
- The `Tbody` that contains any expandable rows must have `isExpanded` logic
- Any `Tr` whose purpsoe is to act as a "control row" (for normal expandable tables) or has the `isControlRow` prop (for compound expansion tables) must have `isExpanded` logic
- Any `Tr` whose purpose is to expand/collapse (the one that actually contains the expandable content) must have `isExpandable` passed to it, in addition to `isExpanded` logic
  - Additionally, any `Td` within these expandable `Tr`'s must wrap any content in an `ExpandableRowContent`

With a basic example structure being:

```js
<Table isExpandable> // pass hasAnimations if opting into animations, or run codemods rule to auto apply it
  ...Thead stuff...
  <Tbody isExpanded={...logic for checking if expanded}>
    <Tr isExpanded={...logic for checking if expanded}> // pass isControlRow for compound expandable implementation as well
      ..."control row" content
    </Tr>
    <Tr isExpandable isExpanded={...logic for checking if expanded}>
      <Td>
        <ExpandableRowContent>
          ...the actual content to display for the cell
        </ExpandableRowContent>
      </Td>
    </Tr>
  </Tbody>
</Table>
```

But this is also a lot of information so let me know what you think.

cc @mcoker did I miss anything/mess anything up in the intended structure here?

    - **Requires opt-in.**
    - **Example:** [Dual list selector with tree](/components/dual-list-selector#with-tree), which has opted in.

- Expandable section 
    - **Example:** All [expandable sections](/components/expandable-section) in our examples and demos.
    - **Note:** When using [detached expandable sections](/components/expandable-section#detached), you will need to use the new `direction` prop to set the direction of the expansion animation.

- Form
    - **Requires opt-in.**
    - **Example:** [Form field groups](/components/forms/form#field-groups), which have opted in.  

- Navigation
    - **Example:** [Expandable navigation examples](/components/navigation#expandable).
    - **Note:** This animation was introduced in our 6.2 release, but performance issues were reported. For this release, we reworked the animation to resolve these issues.

- Search input   
    - **Requires opt-in.**
    - **Example:** [Expandable search input](/components/search-input#with-expandable-button), which has opted in.

- Tree view
    - **Requires opt-in.**
    - **Example:** All [tree view](/components/tree-view) examples and demos, which have opted in. 

> - Table

Author: ```suggestion
- Table (beta)
```

    - **Requires opt-in.**
    - **Example:** [Expandable table](/components/table#expandable) and [compound expandable table](/components/table/#compound-expandable), which have both opted in.
    - **Notes:** 

>         - This release does not include expansion animations for tree tables.
>         - To prevent possible visual issues with [expandable](/components/table#expandable) and [compound expandable tables](/components/table#compound-expandable), ensure that your implementation aligns with our currently recommended structure. In your table's code: 
>             - `<Table>` must have `isExpandable` passed to it.
>             - Any `<Tbody>` containing expandable rows must have `isExpanded` logic.
>             - Any `<Tr>` that acts as a "control row" (for normal expandable tables) or has the `isControlRow` prop (for compound expansion tables) must have `isExpanded` logic.
>             - Any `<Tr>` that expands/collapses and contains the expandable content must have `isExpandable` passed to it, in addition to `isExpanded` logic.
>             - Any `<Td>` within an expandable `<Tr>` must wrap any content in an `ExpandableRowContent`.
>             - The basic structure, which is used in our expandable and compound expandable examples should resemble this: 
> 
>                 ``` 
>                 <Table isExpandable> // if opting into animations, pass the hasAnimations prop or run the enable-animations codemods to auto-apply the prop
>                   ...Thead...
>                   <Tbody isExpanded={...logic for checking for expansion}>
>                     <Tr isExpanded={...logic for checking for expansion}> // for compound expandable, pass isControlRow as well
>                       ..."control row" content...
>                     </Tr>
>                     <Tr isExpandable isExpanded={...logic for checking if expanded...}>
>                       <Td>
>                         <ExpandableRowContent>
>                           ...the content to display for the cell...
>                         </ExpandableRowContent>
>                       </Td>
>                     </Tr>
>                   </Tbody>
>                 </Table>
>                 ```

Reviewer 3: @edonehoo so as of right now the latest updates for Table are in PR in react at [React PR 11927](https://github.com/patternfly/patternfly-react/pull/11927/files). Here are the tweaks we'd need for what you currently have in the release highlights based on the current React changes:

```suggestion
        - This release does not include expansion animations for tree tables.
        - To prevent possible visual issues with [expandable](/components/table#expandable) and [compound expandable tables](/components/table#compound-expandable), ensure that your implementation aligns with our currently recommended structure. In your table's code: 
            - `<Table>` must have `isExpandable` passed to it.
            - Any `<Tbody>` containing expandable rows must have `isExpanded` logic.
            - Any `<Tr>` that acts as a "control row" (for normal expandable tables) or has the `isControlRow` prop (for compound expansion tables) must have `isContentExpanded` logic (this logic should match the `isExpanded` logic being passed to other sub-components).
            - Any `<Tr>` that expands/collapses and contains the expandable content must have `isExpanded` logic (this should always have been the case in order for expansion to work).
            - Any `<Td>` within an expandable `<Tr>` must wrap any content in an `ExpandableRowContent`.
            - The basic structure, which is used in our expandable and compound expandable examples should resemble this: 

                ``` 
                <Table isExpandable> // if opting into animations, pass the hasAnimations prop or run the enable-animations codemods to auto-apply the prop
                  ...Thead content...
                  <Tbody isExpanded={...logic for checking for expansion...}>
                    <Tr isContentExpanded={...logic for checking for expansion...}> // for compound expandable, pass isControlRow as well
                      ..."control row" content...
                    </Tr>
                    <Tr isExpanded={...logic for checking for expansion...}>
                      <Td>
                        <ExpandableRowContent>
                          ...the content to display for the cell...
                        </ExpandableRowContent>
                      </Td>
                    </Tr>
                  </Tbody>
                </Table>
                ```
```

These changes are also required to resolve the row divider bug (where the horizontal divider between rows was visible when a row was expanded, causing a divider between a "control row" and its expanded content - see https://github.com/patternfly/patternfly-react/issues/11812 for context)

Additionally, the React PR I have up for these changes also marks the hasAnimations as beta for Table - all other components with this prop are not marked as beta.

### React 19 support

PatternFly now includes support for React 19 across all components and extensions, enabling you to benefit from  new React features and performance enhancements.

This update is non-breaking, and PatternFly 6 will continue to support React 17 and 18 in line with our support policy. In future releases, we plan to more fully leverage React 19's capabilities to enhance existing PatternFly features and introduce new ones.

### Component animations

As we progress in our effort to integrate motion design across PatternFly components, we added the following new animations in this release. To provide a holistic view of our animated components, we also put together [an interactive animations demo](/design-foundations/motion/react-demos), which we will continue to update as we add additional animation support.   

**Note:** Browsers may slow down or deactivate animations on resource-intensive pages. This can occur when continuous animations (like multiple spinners) are active at the same time as other demanding tasks (like frequent data polling), which can overload the browser's processing capacity.

#### enable-animations codemods

> We have also created an [enable-animations codemod](https://github.com/patternfly/pf-codemods/tree/main/packages/eslint-plugin-pf-codemods/src/rules/v6/enableAnimations), which adds the `hasAnimations` prop to components that require opt-in animations. Keep in mind that, depending on your codebase, this codemod could introduce breaking changes that require further attention.

Reviewer 3: ```suggestion
We have also created an [enable-animations codemod](https://github.com/patternfly/pf-codemods/tree/main/packages/eslint-plugin-pf-codemods/src/rules/v6/enableAnimations), which adds the `hasAnimations` prop to components that require opt-in animations. Keep in mind that, depending on your codebase, this codemod could introduce breaking changes that require further attention. Due to this, when running the codemod you will be asked whether you want to include the Table components for the animations opt-in, or just opt into animations for the react-core package.
```
Since the codemod will now offer a CLI option to include the react-table package as part of the animations optin, wdyt about this tweak?
